SI-6253 HashSet should implement union

rklaehn opened the pull request:
Implements of HashSet.union that reuses the two trees as much as
possible when calculating the union of two sets. This leads to significant
performance improvements as well as to much better structural sharing.

There is a comprehensive correctness test for union since there was not a single test
for HashSet.union before. In addition, there are some tests of the desirable
properties of the new implementation (structural sharing and efficiency
regarding calls of key.hashCode.

The other operations diff and intersect, which are conceptually very
similar to union, are also implemented along with comprehensive test cases
for both correctness and structural sharing.

Note that while it appears that there is some code duplication between the
three methods, they are sufficiently different that it is not possible
to merge them into one without sacrificing performance.

Review by @Ichoran, @axel22

rklaehn commented:
This is not a small change. So I guess I should provide some motivation why it makes sense. 
- The old implementation had basically no structural sharing in some situations. For example, empty union someBigHashSet will rebuild the entire set. The new implementation will always reuse existing nodes when possible.
- Union can be used to efficiently implement ++ if that is a HashSet as well. So any "divide and conquer" algorithm that builds a set recursively from subsets will benefit a lot. Diff can be used to implement -- if that is a HashSet as well.
- Here is a little benchmark for union. It compares the new implementation of union with the old foldLeft-based one by creating sets of different sizes and different overlaps and creating the union. As you can see, the new version is always much faster than the old one. Up to 5 times faster, but never slower. 

```
rklaehn@linux-hmb9:~/projects_git/rklaehn/scala> build/quick/bin/scala -cp ~/projects_git/rklaehn/hashset/lib/Thyme.jar 
Welcome to Scala version 2.11.0-20140102-191511-2fa7cf708e (OpenJDK 64-Bit Server VM, Java 1.7.0_40).
Type in expressions to have them evaluated.
Type :help for more information.

scala> :paste
// Entering paste mode (ctrl-D to finish)


  def unionBench() = {
    import scala.collection.immutable.HashSet
    val th = ichi.bench.Thyme.warmed()

    import th.autoWarmer

    for {
      i <- Seq(1, 10, 100, 1000, 10000)
      o <- Seq(0.0, 0.25, 0.5, 0.75, 1.0)
    } {
      val sep = (i * o).toInt
      val a = HashSet.empty ++ (0 until i)
      val b = HashSet.empty ++ (0 until i).map(_ + sep)
      th.pbenchOffWarm(s"union(0..$i,$sep..${sep + i})")(() => a union b)(() => (a /: b.seq)(_ + _))
    }
  }

// Exiting paste mode, now interpreting.

unionBench: ()Unit

scala> unionBench()
Benchmark comparison (in 5.228 s): union(0..1,0..1)
Significantly different (p ~= 0)
  Time ratio:    2.24573   95% CI 2.23830 - 2.25316   (n=20)
    First     6.570 ns   95% CI 6.559 ns - 6.581 ns
    Second    14.75 ns   95% CI 14.71 ns - 14.80 ns
Benchmark comparison (in 5.253 s): union(0..1,0..1)
Significantly different (p ~= 0)
  Time ratio:    2.25159   95% CI 2.24457 - 2.25862   (n=20)
    First     6.581 ns   95% CI 6.571 ns - 6.591 ns
    Second    14.82 ns   95% CI 14.78 ns - 14.86 ns
Benchmark comparison (in 5.241 s): union(0..1,0..1)
Significantly different (p ~= 0)
  Time ratio:    2.24946   95% CI 2.23710 - 2.26182   (n=20)
    First     6.581 ns   95% CI 6.563 ns - 6.599 ns
    Second    14.80 ns   95% CI 14.73 ns - 14.87 ns
Benchmark comparison (in 6.104 s): union(0..1,0..1)
Significantly different (p ~= 0)
  Time ratio:    1.71989   95% CI 1.71436 - 1.72542   (n=20)
    First     8.578 ns   95% CI 8.568 ns - 8.589 ns
    Second    14.75 ns   95% CI 14.71 ns - 14.80 ns
  Individual benchmarks not fully consistent with head-to-head (p ~= 0)
    First     6.546 ns   95% CI 6.522 ns - 6.569 ns
    Second    14.73 ns   95% CI 14.62 ns - 14.84 ns
Benchmark comparison (in 4.206 s): union(0..1,1..2)
Significantly different (p ~= 0)
  Time ratio:    1.86165   95% CI 1.85407 - 1.86923   (n=20)
    First     16.45 ns   95% CI 16.41 ns - 16.50 ns
    Second    30.63 ns   95% CI 30.54 ns - 30.72 ns
  Individual benchmarks not fully consistent with head-to-head (p ~= 8.780e-07)
    First     16.34 ns   95% CI 16.24 ns - 16.45 ns
    Second    32.44 ns   95% CI 32.03 ns - 32.85 ns
Benchmark comparison (in 4.118 s): union(0..10,0..10)
Significantly different (p ~= 0)
  Time ratio:    1.53700   95% CI 1.52728 - 1.54671   (n=20)
    First     147.4 ns   95% CI 146.8 ns - 147.9 ns
    Second    226.5 ns   95% CI 225.4 ns - 227.6 ns
Benchmark comparison (in 5.742 s): union(0..10,2..12)
Significantly different (p ~= 0)
  Time ratio:    1.68465   95% CI 1.67821 - 1.69110   (n=20)
    First     195.7 ns   95% CI 195.2 ns - 196.2 ns
    Second    329.7 ns   95% CI 328.7 ns - 330.7 ns
Benchmark comparison (in 3.720 s): union(0..10,5..15)
Significantly different (p ~= 0)
  Time ratio:    2.07299   95% CI 2.06175 - 2.08423   (n=20)
    First     226.6 ns   95% CI 225.8 ns - 227.5 ns
    Second    469.8 ns   95% CI 468.1 ns - 471.6 ns
Benchmark comparison (in 4.262 s): union(0..10,7..17)
Significantly different (p ~= 0)
  Time ratio:    2.29537   95% CI 2.28721 - 2.30353   (n=20)
    First     247.6 ns   95% CI 246.9 ns - 248.2 ns
    Second    568.3 ns   95% CI 566.9 ns - 569.6 ns
Benchmark comparison (in 5.183 s): union(0..10,10..20)
Significantly different (p ~= 0)
  Time ratio:    2.57386   95% CI 2.55931 - 2.58842   (n=20)
    First     280.9 ns   95% CI 279.6 ns - 282.1 ns
    Second    723.0 ns   95% CI 720.5 ns - 725.5 ns
Benchmark comparison (in 3.273 s): union(0..100,0..100)
Significantly different (p ~= 0)
  Time ratio:    1.78681   95% CI 1.76898 - 1.80464   (n=20)
    First     1.746 us   95% CI 1.734 us - 1.758 us
    Second    3.120 us   95% CI 3.097 us - 3.143 us
Benchmark comparison (in 4.518 s): union(0..100,25..125)
Significantly different (p ~= 0)
  Time ratio:    2.17185   95% CI 2.13969 - 2.20400   (n=20)
    First     2.159 us   95% CI 2.135 us - 2.182 us
    Second    4.688 us   95% CI 4.641 us - 4.735 us
Benchmark comparison (in 4.016 s): union(0..100,50..150)
Significantly different (p ~= 0)
  Time ratio:    2.54292   95% CI 2.53167 - 2.55417   (n=20)
    First     2.438 us   95% CI 2.433 us - 2.444 us
    Second    6.201 us   95% CI 6.177 us - 6.224 us
Benchmark comparison (in 4.686 s): union(0..100,75..175)
Significantly different (p ~= 0)
  Time ratio:    2.98619   95% CI 2.97008 - 3.00231   (n=20)
    First     2.642 us   95% CI 2.633 us - 2.650 us
    Second    7.889 us   95% CI 7.855 us - 7.923 us
Benchmark comparison (in 5.354 s): union(0..100,100..200)
Significantly different (p ~= 0)
  Time ratio:    3.35433   95% CI 3.34383 - 3.36483   (n=20)
    First     2.862 us   95% CI 2.856 us - 2.867 us
    Second    9.599 us   95% CI 9.576 us - 9.622 us
Benchmark comparison (in 4.033 s): union(0..1000,0..1000)
Significantly different (p ~= 0)
  Time ratio:    1.95758   95% CI 1.94933 - 1.96583   (n=20)
    First     22.08 us   95% CI 22.00 us - 22.16 us
    Second    43.22 us   95% CI 43.14 us - 43.31 us
Benchmark comparison (in 4.807 s): union(0..1000,250..1250)
Significantly different (p ~= 0)
  Time ratio:    3.10782   95% CI 3.09787 - 3.11777   (n=20)
    First     21.68 us   95% CI 21.64 us - 21.72 us
    Second    67.37 us   95% CI 67.20 us - 67.54 us
Benchmark comparison (in 5.871 s): union(0..1000,500..1500)
Significantly different (p ~= 0)
  Time ratio:    4.06475   95% CI 4.05149 - 4.07801   (n=20)
    First     23.11 us   95% CI 23.06 us - 23.16 us
    Second    93.94 us   95% CI 93.72 us - 94.15 us
Benchmark comparison (in 3.546 s): union(0..1000,750..1750)
Significantly different (p ~= 0)
  Time ratio:    4.83324   95% CI 4.78102 - 4.88547   (n=20)
    First     25.43 us   95% CI 25.22 us - 25.64 us
    Second    122.9 us   95% CI 122.1 us - 123.8 us
Benchmark comparison (in 4.016 s): union(0..1000,1000..2000)
Significantly different (p ~= 0)
  Time ratio:    5.20545   95% CI 5.14855 - 5.26236   (n=20)
    First     27.67 us   95% CI 27.43 us - 27.91 us
    Second    144.0 us   95% CI 143.1 us - 145.0 us
Benchmark comparison (in 2.817 s): union(0..10000,0..10000)
Significantly different (p ~= 0)
  Time ratio:    1.96322   95% CI 1.92515 - 2.00129   (n=20)
    First     243.9 us   95% CI 239.7 us - 248.2 us
    Second    478.9 us   95% CI 474.7 us - 483.1 us
Benchmark comparison (in 4.207 s): union(0..10000,2500..12500)
Significantly different (p ~= 0)
  Time ratio:    2.40780   95% CI 2.37765 - 2.43796   (n=20)
    First     335.5 us   95% CI 333.3 us - 337.7 us
    Second    807.8 us   95% CI 799.2 us - 816.5 us
Benchmark comparison (in 5.203 s): union(0..10000,5000..15000)
Significantly different (p ~= 0)
  Time ratio:    2.89014   95% CI 2.86907 - 2.91120   (n=20)
    First     383.4 us   95% CI 381.7 us - 385.0 us
    Second    1.108 ms   95% CI 1.101 ms - 1.115 ms
Benchmark comparison (in 4.384 s): union(0..10000,7500..17500)
Significantly different (p ~= 0)
  Time ratio:    3.49275   95% CI 3.46176 - 3.52375   (n=20)
    First     398.8 us   95% CI 395.7 us - 401.9 us
    Second    1.393 ms   95% CI 1.387 ms - 1.399 ms
Benchmark comparison (in 4.712 s): union(0..10000,10000..20000)
Significantly different (p ~= 0)
  Time ratio:    4.19738   95% CI 4.13619 - 4.25857   (n=20)
    First     390.8 us   95% CI 388.1 us - 393.4 us
    Second    1.640 ms   95% CI 1.619 ms - 1.661 ms
```

rklaehn commented:
For completeness sake, here is diffBench and intersectBench. If anything, the improvements are even larger.

```
scala> :paste
// Entering paste mode (ctrl-D to finish)


  def diffBench() = {
    import scala.collection.immutable.HashSet
    val th = ichi.bench.Thyme.warmed()

    import th.autoWarmer

    for {
      i <- Seq(1, 10, 100, 1000, 10000)
      o <- Seq(0.0, 0.25, 0.5, 0.75, 1.0)
    } {
      val sep = (i * o).toInt
      val a = HashSet.empty ++ (0 until i)
      val b = HashSet.empty ++ (0 until i).map(_ + sep)
      th.pbenchOffWarm(s"diff(0..$i,$sep..${sep + i})")(() => a diff b)(() => (a /: b.seq) (_ - _))
    }
  }

// Exiting paste mode, now interpreting.

diffBench: ()Unit

scala> diffBench()
Benchmark comparison (in 6.619 s): diff(0..1,0..1)
Significantly different (p ~= 0)
  Time ratio:    2.95230   95% CI 2.94527 - 2.95934   (n=20)
    First     7.477 ns   95% CI 7.466 ns - 7.487 ns
    Second    22.07 ns   95% CI 22.03 ns - 22.12 ns
Benchmark comparison (in 6.659 s): diff(0..1,0..1)
Significantly different (p ~= 0)
  Time ratio:    2.96071   95% CI 2.94792 - 2.97349   (n=20)
    First     7.518 ns   95% CI 7.499 ns - 7.538 ns
    Second    22.26 ns   95% CI 22.18 ns - 22.34 ns
Benchmark comparison (in 6.650 s): diff(0..1,0..1)
Significantly different (p ~= 0)
  Time ratio:    2.95003   95% CI 2.93871 - 2.96136   (n=20)
    First     7.524 ns   95% CI 7.507 ns - 7.542 ns
    Second    22.20 ns   95% CI 22.13 ns - 22.27 ns
Benchmark comparison (in 6.661 s): diff(0..1,0..1)
Significantly different (p ~= 0)
  Time ratio:    2.94533   95% CI 2.93519 - 2.95547   (n=20)
    First     7.549 ns   95% CI 7.533 ns - 7.564 ns
    Second    22.23 ns   95% CI 22.17 ns - 22.30 ns
Benchmark comparison (in 6.334 s): diff(0..1,1..2)
Significantly different (p ~= 0)
  Time ratio:    3.42640   95% CI 3.41239 - 3.44041   (n=20)
    First     6.705 ns   95% CI 6.688 ns - 6.723 ns
    Second    22.98 ns   95% CI 22.90 ns - 23.05 ns
Benchmark comparison (in 4.564 s): diff(0..10,0..10)
Significantly different (p ~= 0)
  Time ratio:    3.70357   95% CI 3.68897 - 3.71817   (n=20)
    First     99.32 ns   95% CI 98.97 ns - 99.66 ns
    Second    367.8 ns   95% CI 367.1 ns - 368.5 ns
Benchmark comparison (in 5.242 s): diff(0..10,2..12)
Significantly different (p ~= 0)
  Time ratio:    2.62950   95% CI 2.61752 - 2.64147   (n=20)
    First     140.5 ns   95% CI 140.0 ns - 141.0 ns
    Second    369.5 ns   95% CI 368.4 ns - 370.5 ns
Benchmark comparison (in 4.982 s): diff(0..10,5..15)
Significantly different (p ~= 0)
  Time ratio:    1.64020   95% CI 1.62776 - 1.65264   (n=20)
    First     170.8 ns   95% CI 169.9 ns - 171.6 ns
    Second    280.1 ns   95% CI 278.4 ns - 281.7 ns
Benchmark comparison (in 4.672 s): diff(0..10,7..17)
Significantly different (p ~= 0)
  Time ratio:    1.18467   95% CI 1.17995 - 1.18938   (n=20)
    First     182.8 ns   95% CI 182.4 ns - 183.2 ns
    Second    216.6 ns   95% CI 215.8 ns - 217.3 ns
Benchmark comparison (in 4.838 s): diff(0..10,10..20)
Significantly different (p ~= 0)
  Time ratio:    0.67229   95% CI 0.67005 - 0.67454   (n=20)
    First     182.3 ns   95% CI 181.9 ns - 182.6 ns
    Second    122.6 ns   95% CI 122.2 ns - 122.9 ns
Benchmark comparison (in 3.951 s): diff(0..100,0..100)
Significantly different (p ~= 0)
  Time ratio:    5.93589   95% CI 5.91010 - 5.96168   (n=20)
    First     1.130 us   95% CI 1.127 us - 1.133 us
    Second    6.705 us   95% CI 6.682 us - 6.729 us
  Individual benchmarks not fully consistent with head-to-head (p ~= 0.0096)
    First     1.151 us   95% CI 1.144 us - 1.158 us
    Second    6.637 us   95% CI 6.592 us - 6.682 us
Benchmark comparison (in 3.668 s): diff(0..100,25..125)
Significantly different (p ~= 0)
  Time ratio:    3.96215   95% CI 3.94829 - 3.97601   (n=20)
    First     1.836 us   95% CI 1.832 us - 1.841 us
    Second    7.275 us   95% CI 7.257 us - 7.293 us
Benchmark comparison (in 4.932 s): diff(0..100,50..150)
Significantly different (p ~= 0)
  Time ratio:    2.69667   95% CI 2.68474 - 2.70860   (n=20)
    First     2.089 us   95% CI 2.082 us - 2.096 us
    Second    5.633 us   95% CI 5.618 us - 5.648 us
Benchmark comparison (in 4.014 s): diff(0..100,75..175)
Significantly different (p ~= 0)
  Time ratio:    2.09166   95% CI 2.08341 - 2.09992   (n=20)
    First     1.957 us   95% CI 1.952 us - 1.963 us
    Second    4.094 us   95% CI 4.082 us - 4.105 us
Benchmark comparison (in 4.461 s): diff(0..100,100..200)
Significantly different (p ~= 0)
  Time ratio:    1.61879   95% CI 1.61360 - 1.62397   (n=20)
    First     1.720 us   95% CI 1.716 us - 1.725 us
    Second    2.785 us   95% CI 2.780 us - 2.789 us
Benchmark comparison (in 3.343 s): diff(0..1000,0..1000)
Significantly different (p ~= 0)
  Time ratio:    6.62855   95% CI 6.60384 - 6.65327   (n=20)
    First     20.03 us   95% CI 19.96 us - 20.09 us
    Second    132.7 us   95% CI 132.5 us - 133.0 us
Benchmark comparison (in 4.479 s): diff(0..1000,250..1250)
Significantly different (p ~= 0)
  Time ratio:    5.24240   95% CI 5.22754 - 5.25727   (n=20)
    First     21.57 us   95% CI 21.54 us - 21.61 us
    Second    113.1 us   95% CI 112.8 us - 113.4 us
Benchmark comparison (in 5.448 s): diff(0..1000,500..1500)
Significantly different (p ~= 0)
  Time ratio:    4.69517   95% CI 4.67471 - 4.71563   (n=20)
    First     19.99 us   95% CI 19.92 us - 20.06 us
    Second    93.85 us   95% CI 93.59 us - 94.12 us
Benchmark comparison (in 4.343 s): diff(0..1000,750..1750)
Significantly different (p ~= 0)
  Time ratio:    3.73427   95% CI 3.71230 - 3.75625   (n=20)
    First     17.90 us   95% CI 17.82 us - 17.97 us
    Second    66.83 us   95% CI 66.54 us - 67.12 us
Benchmark comparison (in 3.933 s): diff(0..1000,1000..2000)
Significantly different (p ~= 0)
  Time ratio:    2.90561   95% CI 2.89417 - 2.91705   (n=20)
    First     12.95 us   95% CI 12.91 us - 12.99 us
    Second    37.63 us   95% CI 37.55 us - 37.72 us
Benchmark comparison (in 4.596 s): diff(0..10000,0..10000)
Significantly different (p ~= 0)
  Time ratio:    7.21870   95% CI 7.19076 - 7.24665   (n=20)
    First     209.8 us   95% CI 209.1 us - 210.5 us
    Second    1.515 ms   95% CI 1.512 ms - 1.518 ms
Benchmark comparison (in 4.946 s): diff(0..10000,2500..12500)
Significantly different (p ~= 0)
  Time ratio:    4.49269   95% CI 4.47738 - 4.50800   (n=20)
    First     294.4 us   95% CI 293.6 us - 295.1 us
    Second    1.323 ms   95% CI 1.320 ms - 1.326 ms
Benchmark comparison (in 4.768 s): diff(0..10000,5000..15000)
Significantly different (p ~= 0)
  Time ratio:    3.13382   95% CI 3.11828 - 3.14936   (n=20)
    First     341.0 us   95% CI 340.0 us - 342.1 us
    Second    1.069 ms   95% CI 1.065 ms - 1.073 ms
Benchmark comparison (in 4.260 s): diff(0..10000,7500..17500)
Significantly different (p ~= 0)
  Time ratio:    2.30254   95% CI 2.27458 - 2.33049   (n=20)
    First     346.4 us   95% CI 344.3 us - 348.5 us
    Second    797.7 us   95% CI 789.3 us - 806.1 us
Benchmark comparison (in 3.714 s): diff(0..10000,10000..20000)
Significantly different (p ~= 0)
  Time ratio:    1.58088   95% CI 1.56434 - 1.59742   (n=20)
    First     269.4 us   95% CI 267.7 us - 271.2 us
    Second    425.9 us   95% CI 422.5 us - 429.4 us

scala> :paste
// Entering paste mode (ctrl-D to finish)


  def intersectBench() = {
    import scala.collection.immutable.HashSet
    val th = ichi.bench.Thyme.warmed()

    import th.autoWarmer

    for {
      i <- Seq(1, 10, 100, 1000, 10000)
      o <- Seq(0.0, 0.25, 0.5, 0.75, 1.0)
    } {
      val sep = (i * o).toInt
      val a = HashSet.empty ++ (0 until i)
      val b = HashSet.empty ++ (0 until i).map(_ + sep)
      // this comparison relies on that filter still uses the old builder-based approach
      th.pbenchOffWarm(s"diff(0..$i,$sep..${sep + i})")(() => a intersect b)(() => a filter b)
    }
  }

// Exiting paste mode, now interpreting.

intersectBench: ()Unit

scala> intersectBench()
Benchmark comparison (in 5.512 s): diff(0..1,0..1)
Significantly different (p ~= 0)
  Time ratio:    6.98509   95% CI 6.91759 - 7.05259   (n=20)
    First     7.562 ns   95% CI 7.533 ns - 7.591 ns
    Second    52.82 ns   95% CI 52.35 ns - 53.29 ns
Benchmark comparison (in 5.373 s): diff(0..1,0..1)
Significantly different (p ~= 0)
  Time ratio:    7.09127   95% CI 7.01461 - 7.16793   (n=20)
    First     7.306 ns   95% CI 7.274 ns - 7.338 ns
    Second    51.81 ns   95% CI 51.30 ns - 52.32 ns
Benchmark comparison (in 5.362 s): diff(0..1,0..1)
Significantly different (p ~= 0)
  Time ratio:    7.07818   95% CI 7.01733 - 7.13903   (n=20)
    First     7.318 ns   95% CI 7.292 ns - 7.343 ns
    Second    51.80 ns   95% CI 51.39 ns - 52.20 ns
Benchmark comparison (in 5.339 s): diff(0..1,0..1)
Significantly different (p ~= 0)
  Time ratio:    7.05226   95% CI 6.99623 - 7.10830   (n=20)
    First     7.315 ns   95% CI 7.292 ns - 7.339 ns
    Second    51.59 ns   95% CI 51.22 ns - 51.97 ns
Benchmark comparison (in 5.554 s): diff(0..1,1..2)
Significantly different (p ~= 0)
  Time ratio:    5.66637   95% CI 5.62581 - 5.70693   (n=20)
    First     6.388 ns   95% CI 6.361 ns - 6.414 ns
    Second    36.19 ns   95% CI 35.98 ns - 36.41 ns
Benchmark comparison (in 3.769 s): diff(0..10,0..10)
Significantly different (p ~= 0)
  Time ratio:    7.23175   95% CI 7.17948 - 7.28403   (n=20)
    First     123.7 ns   95% CI 123.1 ns - 124.3 ns
    Second    894.9 ns   95% CI 890.1 ns - 899.7 ns
Benchmark comparison (in 4.065 s): diff(0..10,2..12)
Significantly different (p ~= 0)
  Time ratio:    5.07668   95% CI 5.02680 - 5.12657   (n=20)
    First     154.8 ns   95% CI 153.9 ns - 155.6 ns
    Second    785.7 ns   95% CI 779.1 ns - 792.2 ns
Benchmark comparison (in 4.969 s): diff(0..10,5..15)
Significantly different (p ~= 0)
  Time ratio:    3.92977   95% CI 3.89521 - 3.96432   (n=20)
    First     156.8 ns   95% CI 155.8 ns - 157.7 ns
    Second    616.1 ns   95% CI 612.3 ns - 620.0 ns
Benchmark comparison (in 4.490 s): diff(0..10,7..17)
Significantly different (p ~= 0)
  Time ratio:    3.19404   95% CI 3.16697 - 3.22112   (n=20)
    First     156.7 ns   95% CI 155.8 ns - 157.5 ns
    Second    500.4 ns   95% CI 497.1 ns - 503.7 ns
Benchmark comparison (in 4.469 s): diff(0..10,10..20)
Significantly different (p ~= 0)
  Time ratio:    11.65820   95% CI 11.57416 - 11.74223   (n=20)
    First     28.10 ns   95% CI 27.93 ns - 28.26 ns
    Second    327.6 ns   95% CI 326.2 ns - 328.9 ns
Benchmark comparison (in 3.267 s): diff(0..100,0..100)
Significantly different (p ~= 0)
  Time ratio:    7.79275   95% CI 7.74848 - 7.83701   (n=20)
    First     1.660 us   95% CI 1.653 us - 1.666 us
    Second    12.93 us   95% CI 12.88 us - 12.99 us
Benchmark comparison (in 4.486 s): diff(0..100,25..125)
Significantly different (p ~= 0)
  Time ratio:    5.66249   95% CI 5.63499 - 5.68999   (n=20)
    First     1.858 us   95% CI 1.851 us - 1.866 us
    Second    10.52 us   95% CI 10.49 us - 10.55 us
Benchmark comparison (in 3.685 s): diff(0..100,50..150)
Significantly different (p ~= 0)
  Time ratio:    4.57594   95% CI 4.54748 - 4.60439   (n=20)
    First     1.715 us   95% CI 1.707 us - 1.723 us
    Second    7.846 us   95% CI 7.814 us - 7.878 us
Benchmark comparison (in 4.013 s): diff(0..100,75..175)
Significantly different (p ~= 0)
  Time ratio:    4.91454   95% CI 4.85766 - 4.97143   (n=20)
    First     1.249 us   95% CI 1.241 us - 1.256 us
    Second    6.138 us   95% CI 6.077 us - 6.198 us
Benchmark comparison (in 4.766 s): diff(0..100,100..200)
Significantly different (p ~= 0)
  Time ratio:    5.99767   95% CI 5.93304 - 6.06230   (n=20)
    First     687.7 ns   95% CI 683.2 ns - 692.1 ns
    Second    4.125 us   95% CI 4.089 us - 4.160 us
Benchmark comparison (in 4.260 s): diff(0..1000,0..1000)
Significantly different (p ~= 0)
  Time ratio:    8.70294   95% CI 8.61328 - 8.79261   (n=20)
    First     21.48 us   95% CI 21.28 us - 21.68 us
    Second    187.0 us   95% CI 186.2 us - 187.8 us
Benchmark comparison (in 4.745 s): diff(0..1000,250..1250)
Significantly different (p ~= 0)
  Time ratio:    7.77723   95% CI 7.73948 - 7.81498   (n=20)
    First     19.32 us   95% CI 19.26 us - 19.39 us
    Second    150.3 us   95% CI 149.8 us - 150.8 us
Benchmark comparison (in 3.911 s): diff(0..1000,500..1500)
Significantly different (p ~= 0)
  Time ratio:    7.11103   95% CI 7.01987 - 7.20219   (n=20)
    First     16.68 us   95% CI 16.53 us - 16.82 us
    Second    118.6 us   95% CI 117.4 us - 119.7 us
Benchmark comparison (in 4.376 s): diff(0..1000,750..1750)
Significantly different (p ~= 0)
  Time ratio:    6.09414   95% CI 6.06115 - 6.12714   (n=20)
    First     13.31 us   95% CI 13.25 us - 13.37 us
    Second    81.11 us   95% CI 80.87 us - 81.35 us
Benchmark comparison (in 4.854 s): diff(0..1000,1000..2000)
Significantly different (p ~= 0)
  Time ratio:    5.63285   95% CI 5.59794 - 5.66777   (n=20)
    First     8.187 us   95% CI 8.146 us - 8.229 us
    Second    46.12 us   95% CI 45.95 us - 46.28 us
Benchmark comparison (in 12.63 s): diff(0..10000,0..10000)
Significantly different (p ~= 0)
  Time ratio:    9.22899   95% CI 8.87746 - 9.58052   (n=30)
    First     240.5 us   95% CI 233.5 us - 247.4 us
    Second    2.219 ms   95% CI 2.164 ms - 2.275 ms
Benchmark comparison (in 4.000 s): diff(0..10000,2500..12500)
Significantly different (p ~= 0)
  Time ratio:    6.10580   95% CI 5.95144 - 6.26016   (n=20)
    First     290.1 us   95% CI 285.7 us - 294.6 us
    Second    1.771 ms   95% CI 1.736 ms - 1.807 ms
Benchmark comparison (in 5.104 s): diff(0..10000,5000..15000)
Significantly different (p ~= 0)
  Time ratio:    4.29776   95% CI 4.14361 - 4.45190   (n=20)
    First     315.5 us   95% CI 307.2 us - 323.8 us
    Second    1.356 ms   95% CI 1.323 ms - 1.389 ms
Benchmark comparison (in 3.980 s): diff(0..10000,7500..17500)
Significantly different (p ~= 0)
  Time ratio:    3.28191   95% CI 3.26204 - 3.30177   (n=20)
    First     281.0 us   95% CI 279.9 us - 282.1 us
    Second    922.2 us   95% CI 917.9 us - 926.5 us
Benchmark comparison (in 4.515 s): diff(0..10000,10000..20000)
Significantly different (p ~= 0)
  Time ratio:    3.33861   95% CI 3.31161 - 3.36561   (n=20)
    First     160.6 us   95% CI 159.7 us - 161.4 us
    Second    536.1 us   95% CI 532.8 us - 539.4 us
```

axel22 commented on src/library/scala/collection/immutable/HashSet.scala:
```diff
@@ -128,9 +201,13 @@ object HashSet extends ImmutableSetFactory[HashSet] {
     }
   }
 
+  sealed abstract class LeafHashSet[A] extends HashSet[A] {
```
This should have a ScalaDoc explaining the reason for this new class. Is it because the planned `HashSet2`, `HashSet3`, etc.? If so, maybe this should come once they are implemented.

rklaehn replied:
This is just so you can treat HashSet1 and HashSetCollision1 the same, because they are both leaves as far as the Trie data structure is concerned. I will add a comment.

Regarding HashSet2, HashSet3, ...: both @Ichoran and me agreed that those are a bad idea because of megamorphic call sites and general increase in complexity. So the comments regarding HashSet2, .. should be removed. I did a pull request for that. https://github.com/scala/scala/pull/3323

axel22 replied:
Yep, agreed about `HashSetX`!

rklaehn replied:
I should definitely make this thing private[immutable] as well if possible. We don't want to introduce any new public classes...

axel22 commented on src/library/scala/collection/immutable/HashSet.scala:
```diff
@@ -147,15 +224,50 @@ object HashSet extends ImmutableSetFactory[HashSet] {
+        // Just create a branch node containing the two.
+        makeHashTrieSet(this.hash, this, that.hash, that, level)
+      case that: HashSet1[A] =>
+        if (this.key == that.key) {
```
Notice that `immutable.HashMap`s have a `merged` method that helps resolve keys _and_ values:

https://github.com/scala/scala/blob/v2.10.3/src/library/scala/collection/immutable/HashMap.scala#L91

For consistency reasons, we might want to have a `merged` on `HashSet`s too, that takes a merge resolver function that resolves purely based on keys (since there are no values for sets).
And then implement `union` in terms of that.

rklaehn replied:
Merged has to keep track of the order of arguments, whereas union on a set is symmetrical (If two keys are identical according to equals, it should not matter for a set if you add the key from the left side or from the right side). So I think using the code from merged would decrease performance. 

I am aware that there is a large amount of duplication between HashSet and HashMap, but I think this can not really be tackled in time for 2.11, and in any case I do not know a way to remove the duplication without impacting performance.

axel22 replied:
I'm not so sure about the part with `union` and that it does not matter which set. Although there was no doc about the behaviour, previously `Set` implemented `union` with `++` which gives priority to keeping the left hand side elements. Application that rely on this and use objects with non-structural `equals` might break.

Independently of that, I wasn't implying above that this commit should merge `HashSet` and `HashMap` together, just that maybe we should have a `merged` on hash sets too? And `union` implemented in terms of `merged`? I don't think performance differences for immutable hash sets if `union` is implemented in terms of `merged` are critical here (and certainly don't differ in big-oh) - applications that need super-good performance will use mutable hash sets anyway. Wdyt?

rklaehn replied:
If we have to distinguish between elements that are identical according to equals and hashCode, basically all optimizations in the union method are impossible. For example: when a union b has the same size as b, you nevertheless have to merge all elements of b into a since they might be different elements that just happen to be the same according to equals and hashCode, and you have to keep the original elements of a.

If that is the case, the current implementation of union that adds all elements sequentially is probably the best you can do. 

But using sets to store objects without meaningful equality not just relies on undocumented behavior, but is also pretty strange. If you define a class where some fields do not contribute to equality, you should expect that you can't use such a class as a key in a set. Note that classes with reference equality are not a problem. You will only ever get a problem when you define equality and then exclude some fields on purpose.

By the way: If we have to preserve undocumented behavior, we can not really optimize anything at all since somebody might rely on the fact that the result of empty union largeHashSet is ne largeHashSet, or x.filter(_ => true) ne x. You might as well freeze the entire collections library and only allow bugfixes then.

Regarding performance: I think you should not have to fall back to mutable data structures for good performance. Even outside purely functional programming there are lots of use cases where you want performance _and_ immutability: for example an actor that keeps some set of ids as its internal state and wants to be able to hand it out without performance penalty of making a copy.

And with the improvements for bulk operations scala.collections.immutable.HashSet should be very competitive with the mutable version in many scenarios. In many situations it will be _faster_ than the mutable version because of structural sharing. 

Regarding the code duplication of HashSet and HashMap: the obvious way to get rid of that would be to have the Hash-Trie data structure as a separate class, and then wrap it in two different ways, once as a Set and once as a Map. Of course it has the overhead of the wrapper object, but that only matters for very small sets/maps. If we could make the whole thing more efficient in another place, maybe we could justify the additional indirection.

To my surprise this is already being done in the scala collections with TreeSet, where the RedBlackTree is a separate class that is wrapped in two ways in TreeSet and TreeMap. I think this would be a good way forward for the HashTrie as well, but obviously not for 2.11.

axel22 replied:
I don't think that reference equality applies here - point is, most people know that you're not supposed to use `ne` and `eq` on return types of collections to drive the logic of your application, but many people might be relying on the left-hand-side semantics of union when their `equals` is non-structural. This still leaves some optimizations open.

My point about performance above was that most performance-critical applications do a series single updates on a hash table rather than creating copies, so they would not use immutable data structures to begin with. But scala-debate is a better place for such a discussion, not this pr ;)

But, rather than prolonging this discussion further - my original comment was that for consistency reasons, `immutable.HashSet` should have a `merged` too. And if it did, `union` in terms of `merged` would simplify the code.

That's my 2 cents as a reviewer. If you are sure that the performance gain is worth it, then go ahead with the merge. Would be nice to test that too.

Ichoran replied:
Sets are not used only as mathematical entities.  Preserving annotations is potentially important.  In particular, instead of implementing union in terms of ++, one might expect to implement ++ in terms of union.  Although in general I don't support preserving weird undocumented behavior, having either consistent left-preference or right-preference is pretty common.  I worry about introducing bugs if this changes dramatically.

In many cases structural sharing should obviate the need to slow things down any more anyway.  You already have to check if they're equal.  Adding an eq check won't cost much.  (Granted, it will double the logic in some sections.)

Is there time to make such a change or is it too large to do robustly at this point?

rklaehn replied:
I don't think we need a merger function. Just making union consistently prefer the lhs would be ok, right? That would complicate the code a bit, and also lose some structural sharing in case the lhs is small and the rhs is large. Similar things would have to be done with diff and intersect.

Since structural sharing and performance are the main reason for this pr, I would really prefer not having to do this. But if it is necessary to get it committed, it certainly could be done without losing too much performance, at the expense of needing more code.

The main problem is that I am back to work, so I can not really work on this a lot. The next time I can expend some serious effort on this is hopefully at the weekend, and that depends on whether the wife and kids cooperate...

axel22 commented on src/library/scala/collection/immutable/HashSet.scala:
```diff
@@ -147,15 +224,50 @@ object HashSet extends ImmutableSetFactory[HashSet] {
+      case that: HashSetCollision1[A] =>
+        val ks1 = that.ks + key
+        if (ks1.size == that.ks.size) {
+          // could this check be made faster by doing an eq check?
```
I think we cannot rely on this, and this case is so rare anyway that we should never worry about it.

rklaehn replied:
Size is O(N) whereas an eq check is O(1). So it does make a difference for hashsets with lots of collisions, which happen occasionally if you don't have a good hashcode. I had some performance tests with keys that are specially designed for lots of collisions that would benefit from being able to to an eq check.

But I agree that it is probably unwise to rely on the behavior of ListSet without having a test to ensure the behavior, so I will remove the comments for this PR.

Ichoran replied:
The comment should be something like 

// Could be eq check (faster) if we could guarantee that.ks + key
// would return that.ks if it already contains the key

so that we don't lose the effort noting that this would be better, but we say exactly what preconditions are required.

axel22 commented on src/library/scala/collection/immutable/HashSet.scala:
```diff
@@ -147,15 +224,50 @@ object HashSet extends ImmutableSetFactory[HashSet] {
+      that.union0(this, level)
+    }
+
+    override private[immutable] def intersect0(that: HashSet[A], level:Int, buffer:Array[HashSet[A]], offset0:Int): HashSet[A] =
```
Please make sure to be consistent with the spacing after the `:` colon sign.

axel22 commented on src/library/scala/collection/immutable/HashSet.scala:
```diff
@@ -147,15 +224,50 @@ object HashSet extends ImmutableSetFactory[HashSet] {
+
     override def removed0(key: A, hash: Int, level: Int): HashSet[A] =
-      if (hash == this.hash && key == this.key) HashSet.empty[A] else this
+      if (hash == this.hash && key == this.key) null else this
```
Is there a performance difference that you have observed by not returning `empty` here?

rklaehn replied:
For removed0 it does not make a difference. But I am using removed0 from diff, where I rely on getting null instead of the empty node, so removed0 needs to follow the "null means empty" convention as well.

I wish this was not the case. I hate using null just as much as anybody else. But for the new bulk operations (filter, filterNot, diff, intersect etc. ) it makes a large difference to use null instead of empty, especially for the case where the result set is empty or almost empty. If you don't use the null guard value trick, you have some cases where the new implementation is consistently slower than the builder-based one, which would of course be less than optimal.

rklaehn replied:
By the way: I just noticed that RedBlackTree.scala also uses the ugly but efficient null for empty optimization.

axel22 replied:
If there's a good performance reason to do so (of not allocating empty hash sets), then I agree `null` is the way to go. Alternatively, we could maybe have a `HashSet1` singleton and a method to cast to different type parameter instantiations.

rklaehn replied:
It's not an allocation but the fact that calling x.isEmpty is slower than x eq null, I think. One is a virtual method call, the other is just a pointer comparison. Also if you use null for empty you have one less class to deal with and reduce the potential for call sites to become megamorphic.

axel22 commented on src/library/scala/collection/immutable/HashSet.scala:
```diff
@@ -166,15 +278,118 @@ object HashSet extends ImmutableSetFactory[HashSet] {
+      case that: HashSet1[A] =>
+        val ks1 = ks + that.key
+        if (ks1.size == ks.size) {
+          // could this be made faster by doing an eq check?
```
Comparing `size` is fine.

Ichoran replied:
I'd leave a brief version of the comment instead.  // Could use eq check (faster) if ListSet was guaranteed to return itself

axel22 commented on src/library/scala/collection/immutable/HashSet.scala:
```diff
@@ -166,15 +278,118 @@ object HashSet extends ImmutableSetFactory[HashSet] {
+        val ks1 = this.ks ++ that.ks
+        ks1.size match {
+          case size if size == this.ks.size =>
+            // could this check be made faster by doing an eq check?
```
Same as before.

axel22 commented on src/library/scala/collection/immutable/HashSet.scala:
```diff
@@ -243,6 +458,283 @@ object HashSet extends ImmutableSetFactory[HashSet] {
+        }
+        if (rbm == 0) {
+          null
+        } else if(rs == this.size0)
```
Please be consistent with spacing after `if` and indentation of the comments.

rklaehn replied:
I will add a space after if, but what do you mean by indentation of comments?

som-snytt replied:
I noticed the `/**` are Java style instead of like the "this class implements" doc at the top, which is SOS.  (Sanctioned Odersky Style.)

Ichoran commented on src/library/scala/collection/immutable/HashSet.scala:
```diff
@@ -166,15 +278,118 @@ object HashSet extends ImmutableSetFactory[HashSet] {
+        }
+      case that: HashSetCollision1[A] =>
+        val ks1 = this.ks ++ that.ks
+        ks1.size match {
```
I'm not sure the match statement buys us anything over

  val newsize = ks1.size
  if (newsize == this.ks.size) ...

except variation in style between HashSet1 and HashSetCollision1.  In particular, the repeated "size if size" motif seems a little boilerplatey.

Ichoran replied:
Except you do this all over the place, so never mind, may as well leave them.

rklaehn replied:
In some places i switched from an if/else to a match to avoid calling size twice. I took that from looking at your code.

Ichoran commented on src/library/scala/collection/immutable/HashSet.scala:
```diff
@@ -243,6 +458,283 @@ object HashSet extends ImmutableSetFactory[HashSet] {
+    }
+
+    override private[immutable] def union0(that: HashSet[A], level: Int, buffer: Array[HashSet[A]], offset0: Int): HashSet[A] = that match {
+      case that if that eq this =>
```
How much performance does this cost when the sets do not share a common heritage?  If it's non-negligible, it might be good to think more deeply about the relative frequency of encountering sets with common heritage and sets without.

rklaehn replied:
It's an eq check, so it is extremely cheap. The "common heritage" scenario is one that is very important for the use case that caused me to work on this in the first place, so I thought since the whole thing is much faster than before I could justify leaving it in. I have no idea how common this is in usage of HashSet. But note that having good performance on something like this might enable new use cases.

Ichoran commented on src/library/scala/collection/immutable/HashSet.scala:
```diff
@@ -243,6 +458,283 @@ object HashSet extends ImmutableSetFactory[HashSet] {
+      case _ => this
+    }
+
+    override private[immutable] def intersect0(that: HashSet[A], level: Int, buffer: Array[HashSet[A]], offset0: Int): HashSet[A] = that match {
```
How much worse is it to implement intersect in terms of diff?  I.e. (a intersect b) = a diff (b diff a)?  This is a lot of code duplication which is necessary for speed...unless the speed isn't actually needed.

rklaehn replied:
I hate the code duplication as much as you do. But I think a diff (b diff a) would require creating an entire set (b diff a) and throwing it away afterwards, so I would be surprised if it was not significantly slower in some cases.

I think a good long term plan to get rid of duplication is to refactor the HashTrie into a separate structure and wrap it in two different ways, kind of how it is done with RedBlackTree. That would remove a ton of duplication.

Ichoran commented:
Need to rebase and possibly edit by hand to make merging easy.

Ichoran commented:
Still need to rebase and possibly edit by hand, now that I've merged the filter fixes.

rklaehn committed:
SI-6253 HashSet should implement union

Implements of HashSet.union that reuses the two trees as much as
possible when calculating the union of two sets. This leads to significant
performance improvements as well as to much better structural sharing.

There is a comprehensive correctness test for union since there was not a single test
for HashSet.union before. In addition, there are some tests of the desirable
properties of the new implementation (structural sharing and efficiency
regarding calls of key.hashCode).

The other operations diff and intersect, which are conceptually very
similar to union, are also implemented along with comprehensive test cases
for both correctness and structural sharing.

Note that while it appears that there is some code duplication between the
three methods, they are sufficiently different that it is not possible
to merge them into one without sacrificing performance.
